src/main/scala/sample/stream/GroupLogFile.scala

```diff
@@ -29,13 +30,18 @@ object GroupLogFile {
         case other                  => "OTHER"
       }.
       // write lines of each group to a separate file
-      foreach {
+      map {
         case (level, groupFlow) =>
           val output = new PrintWriter(new FileOutputStream(s"target/log-$level.txt"), true)
           // close resource when the group stream is completed
           // foreach returns a future that we can key the close() off of
-          groupFlow.foreach(line => output.println(line)).onComplete(_ => Try(output.close()))
+          val future = groupFlow.foreach(line => output.println(line))
+          future.onComplete { _ => Try(output.close()) }
+          future
       }.
+      buffer(maximumLogLevelCount, OverflowStrategy.error).
+      mapAsync(identity).
+      foreach { _ => Unit }.
```

Comment:
Yes, because it returns a future when all items are iterated; is there a better way to get it?